src/lower/test.rs

```diff
@@ -191,3 +225,126 @@ fn check_parameter_kinds() {
         }
     }
 }
+
+#[test]
+fn two_impls_for_same_type() {
+    lowering_error! {
+        program {
+            trait Foo { }
+            struct Bar { }
+            impl Foo for Bar { }
+            impl Foo for Bar { }
+        }
+        error_msg {
+            "overlapping impls of trait \"Foo\""
+        }
+    }
+}
+
+#[test]
+fn generic_vec_and_specific_vec() {
+    lowering_error! {
+        program {
+            trait Foo { }
+            struct Vec<T> { }
+            struct Bar { }
+            impl Foo for Vec<Bar> { }
+            impl<T> Foo for Vec<T> { }
+        }
+        error_msg {
+            "overlapping impls of trait \"Foo\""
+        }
+    }
+}
+
+#[test]
+fn concrete_impl_and_blanket_impl() {
+    lowering_error! {
+        program {
+            trait Foo { }
+            struct Bar { }
+            impl Foo for Bar { }
+            impl<T> Foo for T { }
+        }
+        error_msg {
+            "overlapping impls of trait \"Foo\""
+        }
+    }
+}
+
+#[test]
+fn two_blanket_impls() {
+    lowering_error! {
+        program {
+            trait Foo { }
+            trait Bar { }
+            trait Baz { }
+            impl<T> Foo for T where T: Bar { }
+            impl<T> Foo for T where T: Baz { }
+            struct Quux { }
+            impl Bar for Quux { }
+            impl Baz for Quux { }
+        }
+        error_msg {
+            "overlapping impls of trait \"Foo\""
+        }
+    }
+}
+
+#[test]
+#[ignore] // Ignored until we have the compat modality
+fn two_blanket_impls_open_ended() {
+    lowering_error! {
+        program {
+            trait Foo { }
+            trait Bar { }
+            trait Baz { }
+            impl<T> Foo for T where T: Bar { }
+            impl<T> Foo for T where T: Baz { }
+        }
+        error_msg {
+            "overlapping impls of trait \"Foo\""
+        }
+    }
+}
+
+#[test]
+fn multiple_nonoverlapping_impls() {
+    lowering_success! {
+        program {
+            trait Foo { }
+            struct Bar { }
+            struct Baz<T> { }
+            impl Foo for Bar { }
+            impl<T> Foo for Baz<T> { }
+        }
+    }
+}
+
+#[test]
+fn local_negative_reasoning_in_coherence() {
+    lowering_success! {
+        program {
+            trait Foo { }
+            trait Bar { }
+            struct Baz { }
+            impl<T> Foo for T where T: Bar { }
+            impl Foo for Baz { }
+        }
+    }
+}
+
+#[test]
+fn multiple_parameters() {
+    lowering_error! {
+        program {
+            trait Foo<T> { }
+            struct Baz { }
+
+            impl<T> Foo<Baz> for T { }
+            impl<T> Foo<T> for Baz { }
+        } error_msg {
+            "overlapping impls of trait \"Foo\""
+        }
+    }
+}
```

Comment:
I think one area where it'd be good to have more test coverage would be associated types.

For example, I imagine something like this ought to be considered overlap:

```rust
trait Foo<T> { }

trait Iterator { type Item; }


struct Vec<T> { }
impl<T> Iterator for Vec<T> { type Item = T; }

impl<T: Iterator> Foo<<T as Iterator>::Item> for T { }
impl<A, B> Foo<A> for B { }
```

Hmm, that's not quite as interesting as I had thought. Well, I feel like there are interesting tests to be had here. =) Let's at least add that one ...